internal/repository/index.go

```diff
@@ -69,10 +73,77 @@ type indexEntry struct {
 	length    uint32
 }
 
+// shardedBlobMap - actually behaves like map[restic.BlobHandle]indexEntry
+type shardedBlobMap [256]struct {
+	// Note that this is not a real ID, but saves the blobType in the first byte!
+	blob map[restic.ID]indexEntry
+}
+
+func newShardedBlobMap() shardedBlobMap {
+	var sm shardedBlobMap
+	for i := range sm {
+		sm[i].blob = make(map[restic.ID]indexEntry)
+	}
+	return sm
+}
+
+// get(h) gives the entry for the given handle h
+// entry, found := sm.get(h) works like entry, found := map[h]
+func (sm shardedBlobMap) get(h restic.BlobHandle) (indexEntry, bool) {
```

Comment:
Please make this take a pointer receiver. If we add locks, this breaks. If we change a few lines, the function may not be inlineable anymore and the whole 4kiB header structure gets copied in every call.

Reply:
Thanks for giving the hint! In fact using a pointer improved speed quite a lot (will add benchmark results below).
This is changed now.